go/vt/vtgate/engine/fk_cascade.go

```diff
@@ -82,77 +97,149 @@ func (fkc *FkCascade) TryExecute(ctx context.Context, vcursor VCursor, bindVars
 	}
 
 	for _, child := range fkc.Children {
-		// We create a bindVariable for each Child
-		// that stores the tuple of columns involved in the fk constraint.
+		// Having non-empty UpdateExprBvNames is an indication that we have an update query with non-literal expressions in it.
+		// We need to run this query differently because we need to run an update for each row we get back from the SELECT.
+		if len(child.NonLiteralInfo) > 0 {
+			err = fkc.executeNonLiteralExprFkChild(ctx, vcursor, bindVars, wantfields, selectionRes, child, false)
+		} else {
+			err = fkc.executeLiteralExprFkChild(ctx, vcursor, bindVars, wantfields, selectionRes, child, false)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// All the children are modified successfully, we can now execute the Parent Primitive.
+	return vcursor.ExecutePrimitive(ctx, fkc.Parent, bindVars, wantfields)
+}
+
+func (fkc *FkCascade) executeLiteralExprFkChild(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, selectionRes *sqltypes.Result, child *FkChild, isStreaming bool) error {
+	// We create a bindVariable that stores the tuple of columns involved in the fk constraint.
+	bv := &querypb.BindVariable{
+		Type: querypb.Type_TUPLE,
+	}
+	for _, row := range selectionRes.Rows {
+		var tupleValues []sqltypes.Value
+
+		for _, colIdx := range child.Cols {
+			tupleValues = append(tupleValues, row[colIdx])
+		}
+		bv.Values = append(bv.Values, sqltypes.TupleToProto(tupleValues))
+	}
+	// Execute the child primitive, and bail out incase of failure.
+	// Since this Primitive is always executed in a transaction, the changes should
+	// be rolled back incase of an error.
+	bindVars[child.BVName] = bv
+	var err error
+	if isStreaming {
+		err = vcursor.StreamExecutePrimitive(ctx, child.Exec, bindVars, wantfields, func(result *sqltypes.Result) error { return nil })
+	} else {
+		_, err = vcursor.ExecutePrimitive(ctx, child.Exec, bindVars, wantfields)
+	}
+	if err != nil {
+		return err
+	}
+	delete(bindVars, child.BVName)
+	return nil
+}
+
+func (fkc *FkCascade) executeNonLiteralExprFkChild(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, selectionRes *sqltypes.Result, child *FkChild, isStreaming bool) error {
+	// For each row in the SELECT we need to run the child primitive.
+	for _, row := range selectionRes.Rows {
+		// First we check if any of the columns is being updated at all.
+		skipRow := true
+		for _, info := range child.NonLiteralInfo {
+			// We use a null-safe comparison, so the value is guaranteed to be not null.
+			// We check if the column has updated or not.
+			isUnchanged, err := row[info.CompExprCol].ToBool()
```

Comment:
`ToBool()` is a little fragile. It's probably safe to do here, but it would be nice to have a method that is as forgiving as the mysql engine is

Reply:
The comparison expression is built by us in the planning phase, which is a null-safe-equality check. It is guaranteed to return 0 or 1 in the output, so the `ToBool()` conversion is safe here.